Improve tls certificate documentation (#644)
osism/osism.github.io: docs/guides/configuration-guide/loadbalancer.md

# Loadbalancer

The settings of the following section rely on the mechanisms of Kolla-Ansible, 

> therefore it's a good idea to use the [upstream documenation](https://docs.openstack.org/kolla-ansible/latest/admin/tls.html) 

Reviewer: ```suggestion
therefore it's a good idea to use the [upstream documentation](https://docs.openstack.org/kolla-ansible/latest/admin/tls.html) 
```

## TLS certificates

## General procedure

:::warning
It is very important that important decisions are made at the beginning in connection with
TLS encryption and domains, as these decisions are very difficult to change at a later

> date or can only be changed at great expense.

Reviewer: I think this is a good warning, but we should follow up with which decisions exactly are important and why they can't be changed easily later on.

Without this information the reader is a little bit lost imho.

Author: You're right, the sentence is very generic.
I've tried to describe it a little differently - maybe it's better this way.

## TLS certificates

:::warning
To avoid unnecessary additional work and problems, it is recommended that you configure TLS with the intended target

> configuration of the specific environment during the initial rollout process.

Reviewer: This is very vague, without the knowledge what can go wrong I don't think any user with deep knowledge about this topic can understand this warning.

What work? What problems? Imho a warning should be very specific what can go wrong why.

:::

For example: Changes to the configuration of TLS or the fully qualified domain names (FQDNs) will result in new URLs (with and without the https prefix)

> stored in the Openstack databases, and in the case of self-signed certificates, the CA certificate must be distributed to all participating clients.

Reviewer: This doesn't explain anything really.

When using self-signed certificates, you always need to distribute/update the CA if anything changes?

Which specific changes to the TLS configuration results in new URLs? Even as a maintainer of this software I would have to look this up, so it should be explained to the operator configuring this, because they certainly won't know.

Please add something why this is a specific problem, and if it is a problem, how it can be fixed (e.g. new URLs in the DB), or if it can't be fixed easily, just say so.

If we warn about something the warning must be understandable without some secret knowledge, and it must be actionable: What should the operator do, or what should they avoid doing?

Thanks!

As a result, the involved Ansible Plays must at least be executed in the correct order, and not all Ansible Plays can handle all possible configuration transitions on their own.

> Therefore, in some cases, manual adjustments must be made to the systems.

Reviewer: how could the operator run this in a wrong order? ansible plays in general need to be executed in the correct order, this is not really a specific property of setting up TLS in osism? Also _if_ we mention this at all, what is the correct order to execute the ansible plays? It's not mentioned here?

> For this reason, we recommend that you define domains, obtain certificates, and perform configurations in advance.

Reviewer: I'm not sure I understand this either. "in advance", before doing what next?

Please be specific about the order of things being done, e.g. make a numbered list, if it is deemed important.

e.g. when you use let's encrypt, it's not possible to obtain certificates in advance.

You certainly need to define your used FQDN (domain) before deploying osism, but that's orthogonal to using TLS? What kind of "configuration" are you referring to here?

## General procedure

To enable TLS encryption the following steps are needed.

1. Acticate tls encryption for both endpoints

  * To enable external TLS encryption:

    ```yaml title="environments/kolla/configuration.yml"
    kolla_enable_tls_external: "yes"
    ```
  * To enable internal TLS encryption:

    ```yaml title="environments/kolla/configuration.yml"
    kolla_enable_tls_internal: "yes"
     ```

2. Add the combined server certificate and private key to the following locations in the configuration repository:
  * private key & certificates for `kolla_external_fqdn`: `environments/kolla/certificates/haproxy.pem`
  * private key & certificates for `kolla_internal_fqdn`: `environments/kolla/certificates/haproxy-internal.pem`
3. Encrypt the certificates using ansible vault:
   ```
   make ansible_vault_edit FILE=environments/kolla/certificates/haproxy.pem
   make ansible_vault_edit FILE=environments/kolla/certificates/haproxy-internal.pem
   ```

> 4. Add the changesto the Git repository

Reviewer: ```suggestion
4. Add the changes to the Git repository
```

   ```
   git add environments/kolla/certificates/haproxy.pem \
     environments/kolla/certificates/haproxy-internal.pem \
     environments/kolla/configuration.yml

   git commit -m "Add new certificates" environments/kolla/certificates/haproxy.pem \
     environments/kolla/certificates/haproxy-internal.pem \
     environments/kolla/configuration.yml
   ```
5. Rollout changes
   ```
   osism apply loadbalancer
   ```

### Self-signed certificates

OSISM supports the usage of self-signed certificates with a custom CA i.e if you
are running a test installation or for interim purposes.

Two certificate files are required to use TLS securely with authentication,
which will be provided by your custom Certificate Authority:

* the server certificate with private key
* the CA certificate with any intermediate certificates

> The following procedure describes the prepration procedure for the CA, which is later followed

Reviewer: ```suggestion
The following procedure describes the preparation tasks for the CA, which is later followed
```

> by the general steps described above.

Reviewer: Please be specific to which steps this is referring to. I could not tell as a user.

## TLS certificates

:::warning
To avoid unnecessary additional work and debugging, it is recommended that you configure TLS with the intended target
configuration of the specific environment before executing the initial rollout procedures.
:::

Changes to the configuration of TLS (i.e. enable or disable) or fully qualified domain names (FQDNs) will
result in new URLs (with and without the https prefix).
These addresses are often stored in the Openstack database on initial deployment and cannot
be updated by simply modifying the configuration repository and performing an additional rollout.

In the case of self-signed certificates, the CA certificate must be distributed to all participating
https clients in the correct dependency order and in a manner appropriate to the associated Openstack service.

As a result, at a minimum, the involved Ansible Plays must be run in the appropriate order, and not all Ansible Plays

> are designed to to handle all possible configuration transitions on their own.

Reviewer: nit:
```suggestion
are designed to handle all possible configuration transitions on their own.
```

   ```

3. Use in OpenStack
   * Add the custom CA to the configuration repository in the directory `environments/kolla/certificates/ca` with the same
     name like in step 1
   * Configure the custom CA to be copied to the OpenStack containers
     ```yaml title="environments/manager/configuration.yml"
     kolla_copy_ca_into_containers: "yes"
     openstack_cacert: /etc/ssl/certs/ca-certificates.crt
     ```

4. Import the ca certificate to all nodes so that the custom CA is known everywhere and the self-signed certificates are accepted as valid.
   ```
   osism apply certificates
   ```

> 5. Execute all steps in the general documentation above

Reviewer: This is also unspecific, there is nothing labeled in the above docs as being "general", which steps precisely?

### Generating TLS certificates with Let’s Encrypt

Using Let's encrypt certificates is a good alternative to traditional certificate authorities and
greatly simplifies the administration of TLS certificates.

For a working Let's Encrypt configuration, the API endpoints (configured by `kolla_internal_fqdn` and `kolla_external_fqdn`) 
must be accessible from the internet.

> 1. Acticate Let's Encrypt tls encryption for both endpoints

Reviewer: ```suggestion
1. Activate Let's Encrypt tls encryption for both endpoints
```

   ```

3. Use in OpenStack
   * Add the custom CA to the configuration repository in the directory `environments/kolla/certificates/ca` with the same
     name like in step 1
   * Configure the custom CA to be copied to the OpenStack containers
     ```yaml title="environments/manager/configuration.yml"
     kolla_copy_ca_into_containers: "yes"
     openstack_cacert: /etc/ssl/certs/ca-certificates.crt
     ```

4. Import the ca certificate to all nodes so that the custom CA is known everywhere and the self-signed certificates are accepted as valid.
   ```
   osism apply certificates
   ```

5. Execute all steps in the [general procedure](#general-procedure) above

### Generating TLS certificates with Let’s Encrypt

Using Let's encrypt certificates is a good alternative to traditional certificate authorities and
greatly simplifies the administration of TLS certificates.

For a working Let's Encrypt configuration, the API endpoints (configured by `kolla_internal_fqdn` and `kolla_external_fqdn`) 

> must be accessible from the internet.

Reviewer: This is not necessarily true anymore since you can configure distinct ACME servers for internal or external usage, they only need to be reachable from the control nodes, see:

https://docs.openstack.org/kolla-ansible/latest/admin/tls.html#generating-tls-certificates-with-let-s-encrypt

so you can now use a private ACME-CA-Server in an air gapped environment.
